core/threaded/watchdog.c

```diff
@@ -2,83 +2,173 @@
+    LF_ASSERT(base->reactor_mutex, "reactor-mutex not alloc'ed but has watchdogs.");
+
+    // Grab reactor-mutex and start infinite loop.
     LF_MUTEX_LOCK((lf_mutex_t*)(base->reactor_mutex));
-    instant_t physical_time = lf_time_physical();
-    while (physical_time < watchdog->expiration) {
-        interval_t T = watchdog->expiration - physical_time;
-        LF_MUTEX_UNLOCK((lf_mutex_t*)base->reactor_mutex);
-        lf_sleep(T);
-        LF_MUTEX_LOCK((lf_mutex_t*)(base->reactor_mutex));
-        physical_time = lf_time_physical();
-    }
+    while (true) {
 
-    if (watchdog->expiration != NEVER) {
+        // Step 1: Wait for a timeout to start watching for.
+        
+        // Edge case 1: We have already gotten a signal to terminate.
+        if (watchdog->terminate) {
+            goto terminate;
+        }
+
+        // Edge case 2: We have already received a timeout.
+        if(watchdog->expiration != NEVER) {
+            watchdog_wait(watchdog);
+        } else {
+            // Wait for a signal that we have a timeout to wait for on the cond-var.
+            LF_COND_WAIT(&watchdog->cond);
+            
+            // Check whether we actually got a termination signal.
+            if (watchdog->terminate) {
+                goto terminate;
+            }
+
+            // Finally go wait for that timeout.
+            watchdog_wait(watchdog);
+        }
+
+        // At this point we have returned from the watchdog wait. But it could
+        // be that it was to terminate the watchdog.
+        if (watchdog->terminate) {
+            goto terminate;
+        }
+
+        // It could also be that the watchdog was just stopped
+        if (watchdog->expiration == NEVER) {
+            continue;
+        }
+
+        // Finally. The watchdog actually timed out. Handle it.
+        LF_PRINT_DEBUG("Watchdog %p timed out", (void *) watchdog);
         watchdog_function_t watchdog_func = watchdog->watchdog_function;
         (*watchdog_func)(base);
+        watchdog->expiration = NEVER;
+        watchdog->active = false;
     }
-    watchdog->thread_active = false;
 
-    LF_MUTEX_UNLOCK((lf_mutex_t*)(base->reactor_mutex));
+terminate:
+    // Here the thread terminates. 
+    watchdog->active = false;
+    LF_MUTEX_UNLOCK(base->reactor_mutex);
     return NULL;
 }
 
 void lf_watchdog_start(watchdog_t* watchdog, interval_t additional_timeout) {
-    // Assumes reaction mutex is already held.
-
+    // Assumes reactor mutex is already held.
     self_base_t* base = watchdog->base;
-
+    watchdog->terminate = false;
     watchdog->expiration = base->environment->current_tag.time + watchdog->min_expiration + additional_timeout;
 
-    if (!watchdog->thread_active) {
-        lf_thread_create(&(watchdog->thread_id), _lf_run_watchdog, watchdog);
-        watchdog->thread_active = true;
+    // If the watchdog is inactive, signal it to start waiting.
+    if (!watchdog->active) {
+        LF_COND_SIGNAL(&watchdog->cond);
```

Comment:
I'm confused about this `active` field.  If `active == false`, then the thread has terminated, right? In that case, there is no point in signaling it.

Reply:
The watchdog thread has two states:
1. inactive: It is initialized to this. It means that the thread is running, but it is not waiting for a timeout
2. active: it is waiting for a timeout.  Calling lf_watchdog_stop will move it to inactive state. Calling _lf_watchdog_terminate will terminate the thread. A timeout will cause it to move to inactive also.

This is added to avoid having the watchdog thread terminate after a timeout.

Reply:
So `active == false` means that the thread is running, but it is not waiting on a timeout. This means we should not only set the expiration field, but also signal the condition variable